src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -292,9 +294,8 @@ public async Task ApplyBatchChangeToWorkspaceMaybe_NoLockAsync(bool useAsync, Fu
             {
                 solutionChanges = new SolutionChangeAccumulator(oldSolution);
 
-                // Reset the project update state in case this is a retry (we don't want to save the results of the last attempt).
-                projectUpdateState = _projectUpdateState;
-                projectUpdateState = mutation(solutionChanges, projectUpdateState);
+                // Use the _projectUpdateState here to ensure retries run with the original state.
+                projectUpdateState = mutation(solutionChanges, _projectUpdateState);
```

Comment:
nit: consider calling this `finalProjectUpdateState` or something to indicate it will have the end result.